openstudiocore/src/model/test/ThermalZone_GTest.cpp

```diff
@@ -636,33 +623,138 @@ TEST_F(ModelFixture, ThermalZone_Ports)
   EXPECT_TRUE(exhaustPortList.handle().isNull());
 }
 
-TEST_F(ModelFixture, ThermalZone_ZoneControlContaminantController)
+TEST_F(ModelFixture, ThermalZone_Thermostat)
 {
-  Model m;
-  ThermalZone zone(m);
-
-  ZoneControlContaminantController controller(m);
-  auto result = zone.setZoneControlContaminantController(controller);
-  EXPECT_TRUE(result);
-
-  auto returnValue = zone.zoneControlContaminantController();
-  EXPECT_TRUE(returnValue);
-
-  EXPECT_EQ(controller,returnValue.get());
-
-  zone.resetZoneControlContaminantController();
-  EXPECT_FALSE(zone.zoneControlContaminantController());
-
-  EXPECT_EQ(1u,m.getModelObjects<model::ZoneControlContaminantController>().size());
-
-  controller.remove();
-  EXPECT_TRUE(m.getModelObjects<model::ZoneControlContaminantController>().empty());
+  {
+    Model m;
+
+    ThermalZone zone1(m);
+    ThermostatSetpointDualSetpoint thermostat1(m);
+
+    EXPECT_TRUE(zone1.setThermostat(thermostat1));
+    EXPECT_TRUE(zone1.setThermostat(thermostat1));
+
+    ASSERT_TRUE(thermostat1.thermalZone());
+    EXPECT_EQ(zone1,thermostat1.thermalZone().get());
+
+    auto returnvalue = zone1.thermostat();
+    ASSERT_TRUE(returnvalue);
+    EXPECT_EQ(thermostat1,returnvalue.get());
+
+    ThermalZone zone2(m);
+
+    EXPECT_TRUE(zone2.setThermostat(thermostat1));
+    returnvalue = zone2.thermostat();
+    ASSERT_TRUE(returnvalue);
+    EXPECT_NE(returnvalue.get(),thermostat1);
+    auto thermostats = m.getModelObjects<Thermostat>();
+    EXPECT_EQ(2u,thermostats.size());
+
+    ThermostatSetpointDualSetpoint themostat2(m);
+    EXPECT_TRUE(zone2.setThermostat(themostat2));
+    EXPECT_TRUE(returnvalue->handle().isNull());
+    EXPECT_EQ(2u,m.getModelObjects<Thermostat>().size());
+  }
+
+  {
+    Model m;
+    ThermalZone zone(m);
+    ThermostatSetpointDualSetpoint thermostat(m); 
+    zone.setThermostat(thermostat);
+    auto zone2 = zone.clone().cast<model::ThermalZone>();
+
+    EXPECT_TRUE(zone2.thermostat());
+    EXPECT_EQ(2u,m.getModelObjects<model::Thermostat>().size());
```

Comment:
Huh I guess it is since you are checking 

    EXPECT_TRUE(zone2.thermostat());

maybe just pop in a 

    ASSERT_TRUE(zone1.thermostat());
    ASSERT_TRUE(zone2.thermostat());
    EXPECT_NE(zone.thermostat()->handle(), zone2.thermostat()->handle());